toonz/sources/toonzqt/schematicviewer.cpp

```diff
@@ -432,6 +553,200 @@ void SchematicSceneViewer::showEvent(QShowEvent *se) {
+        m_gestureActive = true;
+      }
+
+      if (changeFlags & QPinchGesture::CenterPointChanged) {
+        QPointF centerDelta = (gesture->centerPoint() * getDevPixRatio()) -
+                              (gesture->lastCenterPoint() * getDevPixRatio());
+        if (centerDelta.manhattanLength() > 1) {
+          // panQt(centerDelta.toPoint());
+        }
+        m_gestureActive = true;
+      }
+    }
+  }
+  e->accept();
+}
+
+void SchematicSceneViewer::touchEvent(QTouchEvent *e, int type) {
+  // qDebug() << "[touchEvent]";
+  if (type == QEvent::TouchBegin) {
+    m_touchActive   = true;
+    m_firstPanPoint = e->touchPoints().at(0).pos();
+    // obtain device type
+    m_touchDevice = e->device()->type();
+  } else if (m_touchActive) {
+    // touchpads must have 2 finger panning for tools and navigation to be
+    // functional on other devices, 1 finger panning is preferred
+    if ((e->touchPoints().count() == 2 &&
+         m_touchDevice == QTouchDevice::TouchPad) ||
+        (e->touchPoints().count() == 1 &&
+         m_touchDevice == QTouchDevice::TouchScreen)) {
+      QTouchEvent::TouchPoint panPoint = e->touchPoints().at(0);
+      if (!m_panning) {
+        QPointF deltaPoint = panPoint.pos() - m_firstPanPoint;
+        // minimize accidental and jerky zooming/rotating during 2 finger
+        // panning
+        if ((deltaPoint.manhattanLength() > 100) && !m_zooming) {
+          m_panning = true;
+        }
+      }
+      if (m_panning) {
+        QPointF centerDelta =
+            (mapToScene(panPoint.pos().toPoint()) * getDevPixRatio()) -
+            (mapToScene(panPoint.lastPos().toPoint()) * getDevPixRatio());
+        panQt(centerDelta);
+      }
+    }
+  }
+  if (type == QEvent::TouchEnd || type == QEvent::TouchCancel) {
+    m_touchActive = false;
+    m_panning     = false;
+  }
+  e->accept();
+}
+
+bool SchematicSceneViewer::event(QEvent *e) {
+  /*
+  switch (e->type()) {
+  case QEvent::TabletPress: {
+    QTabletEvent *te = static_cast<QTabletEvent *>(e);
+    qDebug() << "[event] TabletPress: pointerType(" << te->pointerType()
+             << ") device(" << te->device() << ")";
+  } break;
+  case QEvent::TabletRelease:
+    qDebug() << "[event] TabletRelease";
+    break;
+  case QEvent::TouchBegin:
+    qDebug() << "[event] TouchBegin";
+    break;
+  case QEvent::TouchEnd:
+    qDebug() << "[event] TouchEnd";
+    break;
+  case QEvent::TouchCancel:
+    qDebug() << "[event] TouchCancel";
+    break;
+  case QEvent::MouseButtonPress:
+    qDebug() << "[event] MouseButtonPress";
+    break;
+  case QEvent::MouseButtonDblClick:
+    qDebug() << "[event] MouseButtonDblClick";
+    break;
+  case QEvent::MouseButtonRelease:
+    qDebug() << "[event] MouseButtonRelease";
+    break;
+  case QEvent::Gesture:
+    qDebug() << "[event] Gesture";
+    break;
+  default:
+    break;
+  }
+  */
+
+  if (CommandManager::instance()
+          ->getAction(MI_TouchGestureControl)
+          ->isChecked()) {
```

Comment:
I faced a crash at here when quitting the software. It is because actions may be deleted before closing this widget.
How about modifying the source just like I have done to `sceneviewerevents.cpp` ? Please see [here](https://github.com/opentoonz/opentoonz/blob/master/toonz/sources/toonz/sceneviewerevents.cpp#L1105-L1108).

Reply:
Done. Will be included in my next commit after fixing offset issue.